server/config.go

```diff
@@ -150,3 +167,197 @@ func NewConfig() *Config {
 
 	return c
 }
+
+// ValidateAddrs controls the address fields in the Config object
+// and fills in any blanks.
+// The addresses fields must be guaranteed by the caller to either be
+// completely empty, or have both a host part and a port part
+// separated by a colon. In the latter case either can be empty to
+// indicate it's left unspecified.
+func (cfg *Config) ValidateAddrs(ctx context.Context) error {
+	// TODO (2.0): Deprecate Bind.
+	if cfg.Bind != "" {
+		cfg.ListenAddr = cfg.Bind
+	}
+
+	// Validate the advertise address.
+	advScheme, advHost, advPort, err := validateAdvertiseAddr(ctx, cfg.AdvertiseAddr, cfg.ListenAddr)
+	if err != nil {
+		return errors.Wrapf(err, "validating advertise address")
+	}
+	cfg.AdvertiseAddr = schemeHostPortString(advScheme, advHost, advPort)
+
+	// Validate the listen address.
+	listenScheme, listenHost, listenPort, err := validateListenAddr(ctx, cfg.ListenAddr)
+	if err != nil {
+		return errors.Wrap(err, "validating listen address")
+	}
+	cfg.ListenAddr = schemeHostPortString(listenScheme, listenHost, listenPort)
+
+	return nil
+}
+
+// validateAdvertiseAddr validates and normalizes an address accessible
+// Ensures that if the "host" part is empty, it gets filled in with
+// the configured listen address if any, otherwise it makes a best
+// guess at the outbound IP address.
+// Returns scheme, host, port as strings.
+func validateAdvertiseAddr(ctx context.Context, advAddr, listenAddr string) (string, string, string, error) {
+	listenScheme, listenHost, listenPort, err := getListenAddr(listenAddr)
+	if err != nil {
+		return "", "", "", errors.Wrap(err, "getting listen address")
+	}
+
+	advScheme, advHostPort := splitScheme(advAddr)
+	advHost, advPort := "", ""
+	if advHostPort != "" {
+		var err error
+		advHost, advPort, err = net.SplitHostPort(advHostPort)
+		if err != nil {
+			return "", "", "", errors.Wrapf(err, "splitting host port: %s", advHostPort)
+		}
+	}
+	// If no advertise scheme was specified, use the one from
+	// the listen address.
+	if advScheme == "" {
+		advScheme = listenScheme
+	}
+	// If there was no port number, reuse the one from the listen
+	// address.
+	if advPort == "" || advPort == "0" {
+		advPort = listenPort
+	}
+	// Resolve non-numeric to numeric.
+	portNumber, err := net.DefaultResolver.LookupPort(ctx, "tcp", advPort)
+	if err != nil {
+		return "", "", "", errors.Wrapf(err, "looking up non-numeric port: %v", advPort)
+	}
+	advPort = strconv.Itoa(portNumber)
+
+	// If the advertise host is empty, then we have two cases.
+	if advHost == "" {
+		if listenHost == "0.0.0.0" {
+			advHost = OutboundIP().String()
+		} else {
+			advHost = listenHost
+		}
+	}
+	return advScheme, advHost, advPort, nil
+}
+
+// OutboundIP gets the preferred outbound ip of this machine.
+func OutboundIP() net.IP {
+	conn, err := net.Dial("udp", "8.8.8.8:80")
```

Comment:
what's the idea behind using 8.8.8.8:80 here? I guess Dial for udp doesn't actually contact the remote host, and we want to make sure to use an address which is external so that we have a good chance of getting an interface with an externally routable address. 

I played around with using 192.0.0.8 which is the "IPv4 dummy address", but I don't know that this is actually better - it produced the same result as 8.8.8.8:80 on my machine.

Reply:
I really don't know if there's an ideal way to do this. I tried using `net.Interfaces()`, but that just seems like a random choice as to which IP to return at that point. This method seemed clever, and it's returning the IP that would be used if it actually made a connection to `8.8.8.8`. But with that said, I think that target IP could be any external address. Really the only limitation is that it not be something local like `127.0.0.1`, in which case `Dial` will use a local route. Do you think changing that to `192.0.0.8` is more clear?

Reply:
I don't want anyone scanning the code to think that Pilosa needs to be able to hit 8.8.8.8 in order to work. I actually think that keeping it a known external address is probably likely to work in the most situations. I could see a case where someone screwed up a route entry and 192.0.0.0/8 gets routed to the wrong place.

So I guess let's just add a comment explaining what's going on?

Reply:
+1 on the comment.  My gut says that hack will not work without a real address.  The one test i guess we should try is to what happens when the network is totally private and unable to route to the public addresses.

Reply:
it should work @tgruben `net.Dial` with a "udp" network doesn't actually send any data over the network AFAICT. It just looks up the source IP it would use if it were to send a packet.

Reply:
What if we used the gossip seed IP address? That would ensure it chose the correct network in the case that the NIC used to connect to the internet is not the NIC used for Pilosa. Not sure how hard it is to get that config value from this code.

Reply:
in the case of a single node cluster, that *might* result in the node advertising localhost which the client would then pick up and try to use? I think the 8.8.8.8 approach is probably slightly safer, but neither is going to be perfect.

Reply:
I don't think this configuration is relevant to a single-node cluster. Additionally, if you are configuring a local cluster (multiple processes on one machine), and you are configuring your gossip seed as a port on localhost, wouldn't you want to advertise on localhost?

Reply:
I'm envisioning a situation where someone is standing up a single node server. They set up bind to 0.0.0.0 and don't set up advertise. For some reason, they pass localhost as the only seed. 

When the client connects, it will do a `/status` request, and see only localhost which it will then try to use when doing imports and such.

Reply:
@codysoyland @jaffee keep in mind this lookup only applies if the `bind` address has been specified to be `0.0.0.0`. It will never happen as a default.

Reply:
@codysoyland can you describe the config situation that you're worried about?
Is it that someone would specify:
```
bind: "0.0.0.0"
advertise: ""
gossip-seeds: ["localhost:14000"]
```

Reply:
sorry, @jaffee I didn't see your response before posting this, but i think my config example described what your concern was.

Reply:
Yeah I wasn't actually considering that the client uses the advertise url to optimize direct writes to shards. I'm not sure how that problem should be approached considering it doesn't work when Pilosa is behind a load balancer. Maybe 8.8.8.8 is the best solution but I feel like there has to be a better way.

Reply:
I'm fine with this as is - I think it's significantly better than the current state and shouldn't block merging. I do agree though that we could probably do even better, but I'm not 100% sure what that looks like.